source/material_model/compositing.cc

```diff
@@ -0,0 +1,207 @@
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include <deal.II/base/std_cxx11/array.h>
+#include <aspect/material_model/compositing.h>
+#include <utility>
+#include <limits>
+
+using namespace dealii;
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    // Parse property names
+    template <int dim>
+    Property::MaterialProperty
+    Compositing<dim>::parse_property_name(const std::string &s)
+    {
+      if (Property::property_map.count(s) == 0)
+        AssertThrow (false,
+                     ExcMessage ("The value <" + s + "> for a material "
+                                 "property is not one of the valid values."));
+      return Property::viscosity;
+      return Property::property_map.at(s);
+    }
+
+    //Copy the requested data for one model
+    template <int dim>
+    void
+    Compositing<dim>::copy_required_properties(const unsigned int model_index,
+                                               const typename Interface<dim>::MaterialModelOutputs &base_output,
+                                               typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      if (model_property_map.at(Property::viscosity) == model_index)
+        out.viscosities = base_output.viscosities;
+      if (model_property_map.at(Property::density) == model_index)
+        out.densities = base_output.densities;
+      if (model_property_map.at(Property::thermal_expansion_coefficient) == model_index)
+        out.thermal_expansion_coefficients = base_output.thermal_expansion_coefficients;
+      if (model_property_map.at(Property::specific_heat) == model_index)
+        out.specific_heat = base_output.specific_heat;
+      if (model_property_map.at(Property::thermal_conductivity) == model_index)
+        out.thermal_conductivities = base_output.thermal_conductivities;
+      if (model_property_map.at(Property::compressibility) == model_index)
+        out.compressibilities = base_output.compressibilities;
+      if (model_property_map.at(Property::entropy_derivative_pressure) == model_index)
+        out.entropy_derivative_pressure = base_output.entropy_derivative_pressure;
+      if (model_property_map.at(Property::entropy_derivative_temperature) == model_index)
+        out.entropy_derivative_temperature = base_output.entropy_derivative_temperature;
+      if (model_property_map.at(Property::reaction_terms) == model_index)
+        out.reaction_terms = base_output.reaction_terms;
+    }
+
+    template <int dim>
+    void
+    Compositing<dim>::evaluate(const typename Interface<dim>::MaterialModelInputs &in,
+                               typename Interface<dim>::MaterialModelOutputs &out) const
+    {
+      typename Interface<dim>::MaterialModelOutputs base_output(out.viscosities.size(),
+                                                                this->introspection().n_compositional_fields);
```

Comment:
Ok, I wrote a nice long comment, and then realized I am probably wrong. Here is my original comment: 

The current design of this function is for sure the one that uses the least amount of memory. However, I am a bit concerned about the speed, because you have a lot of if-loops in the `copy_required_properties` that have to be evaluated once per `number_of_models * number_of_properties` and in each call you have to find the element of a (admittedly small) map. What about changing the structure to evaluating all material outputs and storing them in a `std::vector<typename Interface<dim>::MaterialModelOutputs>`. Then you can hand over a reference to this vector to a modified `copy_required_properties` function, that essentially does the following for each property:
```
out.reaction_terms = base_outputs[model_property_map.at(Property::reaction_terms)].reaction_terms;
```
This way you would save all of the ifs, and only have `n_properties` lookups into the map.

After writing all this I realized that `n_properties` and `models.size` is small, and allocating the `std::vector<typename Interface<dim>::MaterialModelOutputs>` is very slow (lots of vectors to allocate), so your algorithm is probably faster. Bottom line: Ignore all of what I wrote above, just sometimes it is worth thinking about the complexity and speed of your algorithms :smile: